### ISSUE TITLE
#9 contradits #6 somewhat

### ISSUE THREAD

**Issue by Mdeuschl** [parsecsv/csv-spec#5]
#9 reads "When a field enclosed in double quotes has spaces before and/or after the double quotes, the spaces MUST be ignored"
But #6 says "Spaces are considered part of a field and MUST NOT be ignored."

In my view this overcomplicates the parsing. This practically means that the file cannot easily be parsed one way. I'd have to remember if the field started with spaces only and then if I find a quote go back to where the first non-space was. Space is not a special character other than for this purpose.

My suggestion:
If the first character after the comma _is_ a  " (double quote) the field should be considered quoted - it spans until the next single ".
If the first character after the comma _is not_ a double quote the field should be interpreted as not quoted thus it spans until the next new line or next comma whichever comes first. If there are double quotes inside the field then treat them as regular characters.
This way I'd never have to go backwards.

Are there many csv exporters that use ", " as field delimiter for quoted fields?

**Comment by jimeh** [comment]
You are right, those two clauses are a little contradictory, but they're meant to build on each other based on their order. Maybe clause 6 could mention that it does not apply to quoted fields.

As for your suggestion, I don't think it will work, as clauses 7 and 8 specify that any field containing a double quote (and other special characters) must be enclosed in double quotes, and a double quote within a quoted field must be escaped by preceding it with another double quote.

I have sadly seen CSV files which have spaces around the field separators while using quoted fields. That's why clause 9 exists :)

All that said, it's been years since I last looked at this, and I don't recall how Excel, Google Sheets, and others deal with double quotes in these kind of situations. I'll find some time to refresh my memory in the next couple of days to double check things still make sense.

**Comment by Mdeuschl** [comment]
I have done a quick comparison of Excel (Mac), Numbers (Mac) and Google sheets. They all don't put spaces before or after the field separator. They all do textbook exports when quotes or field separators are in the fields: single quote at start and finish, doubled quotes to escape
[Excel.csv](https://github.com/parsecsv/csv-spec/files/7783927/Excel.csv)
[Google Sheets.csv](https://github.com/parsecsv/csv-spec/files/7783928/Google.Sheets.csv)
[Numbers.csv](https://github.com/parsecsv/csv-spec/files/7783929/Numbers.csv)

When I try to mess up those exported files by adding a space between the field separators and quoted fields I get these results:
Excel and Numbers treat the field as unquoted, don't un-escape doubled quotes and treat a field separator inside the quoted field as a regular field separator.
Google Sheets ignores the space.

When I replace that space with a all three treat the field as unquoted like Excel and Numbers above.

**Comment by martin-eden** [comment]
> If the first character after the comma _is_ a " (double quote) the field should be considered quoted - it spans until the next single ".
> If the first character after the comma _is not_ a double quote the field should be interpreted as not quoted thus it spans until the next new line or next comma whichever comes first.

+1 for this. I wrote my .csv parser with same logic.

Because sane list-of-lists format allows parser with fixed lookahead. To allow it output results along with getting data from stream.

Not like ` ` , ` ` ` `, *terabyte more spaces* -- we still can not decide what it is: padded quoted field or unquoted field. And probably we will never know as our parser will die from memory exhaustion.
